src/plugins/intel_gna/transformations/convert_unaligned_concat_to_cnn.cpp

```diff
@@ -0,0 +1,534 @@
+    constexpr int64_t newConcatAxis = 1;
+    if (trivialConcatFromStrides(orgConcatNode)) {
+        return ConvertTrivialConcatFromStrides(orgConcatNode);
+    }
+    ngraph::NodeVector allNewComponents;
+    const auto orgConcatOutput = orgConcatNode->output(0);
+    const auto concatOutputType = orgConcatOutput.get_element_type();
+    const auto inputs = orgConcatNode->inputs();
+    auto totalOffset = std::accumulate(inputs.cbegin(), inputs.cend(), static_cast<size_t>(0),
+        [](const size_t& sum, const ov::Input<ov::Node>& i) {return sum + ov::shape_size(i.get_shape()); });
+    auto outputNode = inputs.rbegin()->get_source_output();
+
+    // The loop to convert concat components into GNA-friendly primitieves goes
+    // from the last concat component as the component handling may overwrite
+    // some previous components buffer
+    for (auto input = inputs.rbegin(); input != inputs.rend(); ++input) {
+        const auto sizeCurrent = ov::shape_size(input->get_shape());
+        totalOffset -= sizeCurrent;
+        ngraph::NodeVector components;
+
+        if (totalOffset % elementsAligned == 0) {
+            components = GetComponentsForAlignedComponent(outputNode);
+        } else {
+            components = GetComponentsForNotAlignedComponent(outputNode, concatOutputType, totalOffset);
+        }
+        const auto previousInput = std::next(input);
+        if (previousInput != inputs.rend()) {
+            auto result = EnforceOrder(components, previousInput->get_source_output());
+            components = { result.subgraph };
+            outputNode = result.previousNode;   // for the next iteration
+        }
+
+        allNewComponents.insert(allNewComponents.begin(), components.begin(), components.end());
+    }
+
+    auto newConcat = std::make_shared < ngraph::opset7::Concat>(allNewComponents, newConcatAxis);
+
+    const auto originalConcatOutputShape = orgConcatOutput.get_shape();
+    const auto originalConcatOutputReShape = CreateConstantU64V(originalConcatOutputShape);
+    auto newReshapeOutput = std::make_shared<ngraph::opset7::Reshape>(newConcat, originalConcatOutputReShape, false);
+
+    newConcat->get_rt_info()["newUnalignedConcat"] = rtValue(1);
+
+    for (auto&& inputOfConcatConsumer : orgConcatOutput.get_target_inputs()) {
+        inputOfConcatConsumer.replace_source_output(newReshapeOutput);
+    }
+    newReshapeOutput->set_friendly_name(orgConcatNode->get_friendly_name());
+}
+
+NGRAPH_RTTI_DEFINITION(UnalignedConcatPass, "UnalignedConcatPass", 0);
+UnalignedConcatPass::UnalignedConcatPass() {
+    auto concatPattern = ngraph::pattern::wrap_type<ngraph::opset7::Concat>();
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+        auto& pattern_map = m.get_pattern_value_map();
+        auto concat_node = pattern_map.at(concatPattern).get_node_shared_ptr();
+        auto concat = std::dynamic_pointer_cast<ngraph::opset7::Concat>(concat_node);
```

Comment:
it's better to move checks for concat into the matcher passing the checker as an argument of ngraph::pattern::wrap_type<ngraph::opset7::Concat>()

Reply:
checker added

Reply:
@elilobanova Please resolve if it suits you.